<a href="https://colab.research.google.com/github/MotazSameh/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. My Lane (or Freestyle) and Why

## Selected Lane

I selected **Lane 2: Refresh / Content Opportunity Scoring**.

### Why this lane?

Many websites contain hundreds or thousands of pages, but not all of them need to be updated at the same time. The challenge is deciding which pages should be refreshed first to maximize the impact of the team's effort.

This lane focuses on identifying pages that are likely to be declining in performance and ranking them according to their refresh priority. Instead of manually reviewing every page, a machine learning model can help prioritize the pages that deserve immediate attention.

I chose this lane because it addresses a practical business problem, aligns with the starter pipeline provided by FlyRank, and provides an opportunity to build an interpretable machine learning solution that supports real content decisions.

# 2. The Question

## Research Question

**Which content pages should be prioritized for refresh because they are most likely to be declining in performance?**

### Decision

The decision is determining which pages should be reviewed and refreshed first.

### Action

The SEO or content team will prioritize the highest-ranked pages for content updates instead of reviewing every page manually.

### Cost of a Wrong Recommendation

If the model incorrectly recommends a healthy page, the team may waste time updating content that does not need attention.

If the model fails to identify a truly declining page, that page may continue losing traffic, clicks, and business value before anyone notices the problem.

### Why Machine Learning?

Many factors can influence whether a page is declining, including impressions, CTR, page age, content length, and update history. These relationships are difficult to capture with simple hand-written rules.

A machine learning model can learn patterns from historical data and rank pages according to their likelihood of needing a refresh, helping the team make more informed decisions.

# 3. Quick Look at the Data

The following statistics provide a quick overview of the starter dataset and support my choice of the **Refresh / Content Opportunity Scoring** lane.

Instead of assuming that pages need refreshing, I first explore the data to understand whether declining pages show measurable differences from other pages.

In [3]:
import pandas as pd

df = pd.read_csv("content_refresh_anonymized.csv")

# Create binary label
df["is_declining"] = (df["trend_direction"].str.lower() == "down")

print("=" * 60)
print("Dataset Overview")
print("=" * 60)

print(f"Total pages: {len(df):,}")

declining_rate = df["is_declining"].mean() * 100
print(f"Declining pages: {declining_rate:.2f}%")

print("\nAverage Impressions (90 days)")
print(df.groupby("is_declining")["impressions_90d"].mean().round(2))

print("\nAverage Days Since Last Update")
print(df.groupby("is_declining")["days_since_last_update"].mean().round(2))

Dataset Overview
Total pages: 30,000
Declining pages: 54.21%

Average Impressions (90 days)
is_declining
False    5533.31
True     4919.10
Name: impressions_90d, dtype: float64

Average Days Since Last Update
is_declining
False    42.37
True     49.25
Name: days_since_last_update, dtype: float64


### Observations

From the starter dataset, about **54.21%** of the pages are classified as declining, indicating that content refresh is a common challenge rather than a rare event.

Declining pages receive fewer average impressions (**4,919**) compared to non-declining pages (**5,533**), suggesting that declining content tends to lose visibility over time.

Additionally, declining pages have not been updated for a longer period on average (**49.25 days**) than non-declining pages (**42.37 days**). This supports the idea that update history may be a useful signal when prioritizing pages for content refresh.

These observations suggest that predicting refresh opportunities is a meaningful machine learning problem rather than simply applying a fixed rule.

In [4]:
df.head()

,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct,is_declining
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4,True
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7,True
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9,True
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8,False
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7,True


# 4. Careful Words: What I Can and Can't Claim

### What I can claim

- The exploratory analysis shows patterns in the starter dataset.
- Some features, such as impressions and update history, appear to be associated with declining pages.
- A machine learning model may help prioritize pages for review based on these observable signals.

### What I cannot claim

- I cannot claim that updating a page will automatically improve its ranking or traffic.
- I cannot conclude that one feature directly causes a page to decline.
- The current analysis is exploratory and based on the provided anonymized dataset. Any conclusions should be validated on additional data before being used in production.

# 5. Self-Check

- ✅ I selected a provisional lane.
- ✅ I clearly defined the research question.
- ✅ I identified the business decision and the resulting action.
- ✅ I explained the cost of an incorrect recommendation.
- ✅ I supported my lane choice with real statistics from the starter dataset.
- ✅ I used careful language and avoided making unsupported claims.